add BaseButtonProps

elanalynn opened the pull request:
### Background

Closes: https://github.com/Shopify/argo-private/issues/1487
Relates to: https://github.com/Shopify/argo-rfcs/issues/2

### Solution

This adds the `BaseButtonProps` decided on in this RFC discussion. These props are not yet consumed by either ui-extensions library. 

### 🎩

- This API is not yet consumed and has no impact on the ui-extensions libraries.

lemonmade commented on packages/ui-extensions/src/types/Button.ts:
```diff
@@ -0,0 +1,58 @@
+   * `secondary`: button used for secondary actions not blocking user progress. Ex: Download Shop app
+   * `plain`: Renders a button that visually looks like a Link
+   * @default 'primary'*/
+  kind?: 'primary' | 'secondary' | 'plain';
```
IMO this prop is too design-system specific, there's no guarantee checkout will be able to frame the different button types this way

elanalynn replied:
@lemonmade You liked this approach in April 😂 !  https://github.com/Shopify/argo-rfcs/issues/2#issuecomment-824058184. That said, I know things I things evolve. @jffortier addresses these concerns below. If these appearance is still editable by the merchant, does this sit better with you? On the Admin side, this maps pretty well with Polaris.  While they don't have `secondary`, but we can map this to the non-primary kind.

lemonmade replied:
Hmm well the evidence is pretty overwhelming :P I think it will be pretty difficult to have a shared set of button types that are common across all our apps, but it seems like no one else is super worried about it so 👍

lemonmade replied:
I'd at least remove the `@default` here though — I don't think this is the default we will want within checkout in the end, and in general defaults seem like a pretty host-specific concern.

lemonmade commented on packages/ui-extensions/src/types/Button.ts:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * Destination to link to, renders a Link
+   */
+  to?: string;
```
I feel like there should be some conversation about this one. There are many different options — `url` (as Polaris uses), `to` (as I think Checkout uses), `href`, `link`, ... I think it's also important to consider how this prop could potentially change in the future, including in ways that may diverge between the hosts. In particular, I think it's likely we will want a way of providing a `to` that doesn't hardcode assumptions about URL structure (that is, `to` would not just be an arbitrary path/ URL). This is really important in checkout, where buttons from extensions will need to respect the fact that checkouts can sometimes be on a single page, or spread across multiple pages, depending on the merchant theme.

lemonmade commented on packages/ui-extensions/src/types/Button.ts:
```diff
@@ -0,0 +1,58 @@
+   *  Excludes 'success' | 'warning' since buttons in a Banner will be `monochrome`
+   *  and use the default colors set by the Banner.
+   */
+  appearance?: 'monochrome' | 'critical';
```
IMO this is also an admin design system prop, I also find it hard to understand the difference between this and `kind` (neither prop seems to imply it adds semantic meaning, so they all just seem like visual variations)

jffortier replied:
`kind` in checkout defines the semantic of the button. It does come with some default visual styles, but they are editable by the merchant.

`appearance` is the visual prop. It is combined with the kind prop to offer more flexibility.

This is what is possible with the defaults styles of Checkout.
![image](https://user-images.githubusercontent.com/739905/121959179-6dda1600-cd32-11eb-822e-574ce3bb43ac.png)

lemonmade commented on packages/ui-extensions/src/types/Button.ts:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * Whether the button should fill all available inline space.
+   * */
+  fill?: boolean;
```
We do have this in the button props, but I am not sure it fits with some of the more recent APIs we've done. In a lot of those APIs, we have left ourselves more room by defining `blockSize`/ `inlineSize` props with union types, which I expect we'd use for `Button` too when we do a polish pass on our components (e.g., `<Button inlineSize="fill" />`). I think this impacts `size` as well, which I think will be confusing if we have `blockSize`/ `inlineSize` as well. cc/ @jffortier @jun-shop

lemonmade replied:
I think I'm just pretty hesitant to commit to `boolean` props unless it is clear there will never be more than two options. If we choose `fill` for this, and later want a prop that is able to restrict the max size of the button or something similar to that, we are signed up for having a mix of attributes that do similar, but different, things. In contrast, using `inlineSize="fill"` here gives us tons of flexibility for the future to add more semantic sizes, or to allow for arbitrary sizes.

Don't feel super strongly about this one, but other than `disabled`, I haven't seen many props that are truly boolean in the long run.

jffortier replied:
I'm good with `inlineSize="fill"`. We can make that change in Checkout.

lemonmade commented on packages/ui-extensions/src/types/Button.ts:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * A label replacing the loading indicator when user prefers reduced motion
+   */
+  loadingLabel?: string;
```
This API is also a little awkward when we have a separate `loading` prop, I think we might look at consolidating the label as an optional property on an object you can pass to `loading` in Checkout... cc/ @jffortier

jffortier replied:
TBH, this option is a nice touch, but I've been close to removing it multiple times. I leaned towards the safe side when I introduce this difference for user that prefers reduced motion. The spinner is such a small motion in the overall UI and I couldn't find any validation that it would cause any harms.

I'd be ok to remove it OR the object route is elegant too. Is this what you had in mind Chris?
`loading?: boolean | { status: boolean; accessibilityLabel?: string; }`

lemonmade commented on packages/ui-extensions/src/types/Button.ts:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * Allows the button to submit a form
+   */
+  submit?: boolean;
```
Yeah, this is also one I'd like to revisit in checkout. We've discussed `role`/ `accessibilityRole` props that I feel are a more flexible way of declaring this kind of additional semantic information cc/ @jffortier @danrosenthal @simontaisne

jffortier replied:
Yes, I prefer `role` as well and is consistent with [View](https://github.com/Shopify/checkout-web/blob/master/packages/checkout-web-ui/src/components/View/View.tsx#L11). This is something we'll need in Pressable too (@jun-shop you have `kind?: button | submit | toggle` at the moment)

elanalynn commented:
@lemonmade @jffortier @vividviolet I addressed the comments and removed the props that we all agreed about (`submit`, and `loadingLabel`). Are we okay with moving forward with `to`, `kind`, `appearance`, and `fill`?

lemonmade reviewed (approved):
These are well documented 👍 I have a few nits and concerns still but if they aren't shared, I am happy to have this 🚢 and we can see how it works out.
